packages/cli/internal/pkg/aws/cdk/progress.go

```diff
@@ -41,19 +49,139 @@ func (p ProgressStream) DisplayProgress(description string) error {
 	return nil
 }
 
-func runProgressBar(ctx context.Context, description string) chan ProgressEvent {
+func ShowExecution(progressEvents []ProgressStream) []Result {
+	var progressResults []*Result
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(len(progressEvents))
+
+	for _, progressStream := range progressEvents {
+		progressResult := &Result{}
+		progressResults = append(progressResults, progressResult)
+		go updateResultFromStream(progressStream, progressResult, &waitGroup)
+	}
+
+	waitGroup.Wait()
+
+	var results []Result
+	for _, progressResult := range progressResults {
+		results = append(results, *progressResult)
+	}
+
+	return results
+}
+
+func updateResultFromStream(stream ProgressStream, progressResult *Result, wait *sync.WaitGroup) {
+	var lastEvent ProgressEvent
+
+	for event := range stream {
+		if event.Err != nil {
+			progressResult.Err = event.Err
+		} else {
+			lastEvent = event
+		}
+	}
+
+	progressResult.Outputs = lastEvent.Outputs
+	progressResult.UniqueKey = lastEvent.UniqueKey
+	wait.Done()
+}
+
+func (client Client) DisplayProgressBar(description string, progressEvents []ProgressStream) []Result {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	singleStream, progressResults := combineProgressEvents(progressEvents)
+
+	barReceiver := runProgressBar(ctx, description, len(progressEvents))
+	for event := range singleStream {
+		barReceiver <- event
+	}
+
+	var results []Result
+	for _, progressResult := range progressResults {
+		results = append(results, *progressResult)
+	}
+
+	return results
+}
+
+func combineProgressEvents(progressEventChannels []ProgressStream) (ProgressStream, []*Result) {
+	var results []*Result
+	receiver := make(ProgressStream)
+	var waitGroup sync.WaitGroup
+	waitGroup.Add(len(progressEventChannels))
+
+	for _, channel := range progressEventChannels {
+		progressResult := &Result{}
+		results = append(results, progressResult)
+		go sendDataToReceiverAndUpdateResult(channel, progressResult, &waitGroup, receiver)
```

Comment:
Feels like we have too many layers of interdependent go routines running here. Could we think of a way to simplify here? Perhaps by redefining the data structures around ProgressStream to allow for easier pooling of events across deployments?

Reply:
I looked through this last night and struggled to simplify the pattern here. The problem is that we are executing several commands in different shells that we are reading which require different event streams. To follow this via a single UI mechanism (such as the receiver), we need to combine the different streams together so that the messaging is concise (this is also required for the Progress Bar). As a result, we need these several Go routines. 

It's also why this previously approved PR required them as well: https://github.com/aws/amazon-genomics-cli/pull/114/files
```
	waitGroup.Add(len(channels))
	for _, c := range channels {
		go multiplexFunc(c)
	}

	go func() {
		waitGroup.Wait()
		close(multiplexedChannel)
	}()
```